text/0000-serialize-deserialize.md

```diff
@@ -0,0 +1,174 @@
+* Start Date: 2018-08-01
+* RFC PR: https://github.com/fusionjs/rfcs/pull/8
+
+# Summary
+
+This RFC describes a process for standardizing on an API for serializing and deserializing state between server and browser plugins.
+
+# Basic example
+
+```js
+import {SharedStateToken, createPlugin} from 'fusion-core';
+
+export default createPlugin({
+  deps: {
+    SharedState: SharedStateToken,
+  },
+  middleware: ({SharedState}) => {
+    const {serialize, deserialize} = SharedState('some-id');
```

Comment:
Do we need to specify a string id here? Wondering if we could generate one instead. Maybe even use a Token from createToken() that could specify the shape of the data?

Reply:
I think we need a string, especially to make the mocking experience good

Reply:
I think it would be worth adding a strawman example of what the testing experience would look like using tokens instead of string ids to the alternatives section. I agree mocking will probably be easier than working with tokens, but maybe it's possible to have a token-based API that's still pretty good?

I'd feel sad adding string ids as part of this API especially considering how much effort we put into making a DI system that didn't rely on them.

Reply:
We could have the id be configured via a token, I just feel like it is kinda overkill. The plugin can export a UUID if you really want it to be guaranteed unique. FWIW, we are using string ids today.

Here is an example of what it would look like configuring the string as a token:

```js
const MyPluginSharedStateIDToken = createToken('MyPluginSharedState');
createPlugin({
  deps: {
    SharedState: SharedStateToken,
    SharedStateID: MyPluginSharedStateIDToken
  },
  middleware({SharedState, SharedStateID}) {
     const {serialize, deserialize} = SharedState(SharedStateID);
  }
});

// ... in main.js
app.register(MyPluginSharedStateIDToken, 'some-id');
app.register(SharedStateToken, MockSharedState({
  'some-id': mockdata
});
```

This works, but honestly I think it is too much boilerplate as the number of apps which need serialization grows. I would prefer a pattern where the plugin exports the serialization id.

```js
// In fusion-plugin-redux-react
export const SerializationID = '@fusion-plugin-redux-react/serialization-id'

// In test code
import {SerializationID} from 'fusion-plugin-redux-react';
app.register(SharedStateToken, MockSharedState({
  [SerializationID]: mockdata
});
```

Reply:
String IDs just feel off to me when we have tokens. Tokens feel like the right tool to me here to avoid naming collisions. Are we not concerned about possible naming collisions at all?

Here's one example of an API I was playing with, still on the fence, but it's using tokens, thoughts?
```js
// ... in main.js
app.enhance(SharedStateToken, SharedState => SharedState(MyPluginSharedStateToken, shareddata));

// In test code
app.enhance(SharedStateToken, SharedState => SharedState(MyPluginSharedStateToken, mockdata));

// In plugin:
const {serialize, deserialize} = SharedState.extract(MyPluginSharedStateToken);
```

Reply:
I think it is pretty safe to use strings for this type of thing, especially since we are doing it today already with no problem. The string could be prefixed with the name of the npm package for example.